enterprise/server/workflow/service/service.go

```diff
@@ -794,6 +798,153 @@ func (ws *workflowService) waitForWorkflowInvocationCreated(ctx context.Context,
 	}
 }
 
+func (ws *workflowService) buildActionHistoryQuery(ctx context.Context, repoUrl string, pattern string, timeLimitMicros int64) (*query_builder.Query, error) {
+	q := query_builder.NewQuery(`SELECT repo_url, pattern, invocation_id, commit_sha, branch_name, created_at_usec, updated_at_usec, duration_usec, invocation_status, success FROM Invocations`)
+	if err := perms.AddPermissionsCheckToQuery(ctx, ws.env, q); err != nil {
+		return nil, err
+	}
+	authenticatedUser, err := ws.env.GetAuthenticator().AuthenticatedUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	q.AddWhereClause("repo_url = ?", repoUrl)
+	q.AddWhereClause("pattern = ?", pattern)
+	q.AddWhereClause("branch_name IN ?", []string{"master", "main"})
+	q.AddWhereClause("role = ?", "CI_RUNNER")
+	q.AddWhereClause("group_id = ?", authenticatedUser.GetGroupID())
+	q.AddWhereClause("updated_at_usec > ?", timeLimitMicros)
+	q.SetOrderBy("created_at_usec", false)
+	q.SetLimit(30)
+
+	return q, nil
+}
+
+func (ws *workflowService) GetWorkflowHistory(ctx context.Context, req *wfpb.GetWorkflowHistoryRequest) (*wfpb.GetWorkflowHistoryResponse, error) {
```

Comment:
it's not that i don't want to split this function up, it's that each phase has like 3 things going on, but i'll figure something out if folks really want me to.